### PR TITLE
RLE homework

### PR DESCRIPTION
**Pull request by Lauren34**

### CODE REVIEW

File: src/main/java/lesser/gameoflife/GameOfLifeComponent.java [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 public class GameOfLifeComponent extends JComponent {
 
-    private static  GameOfLife game;
+    private static GameOfLife game;
     private static int cellSize = 20;
```
Comment (andrewoid) [inline src/main/java/lesser/gameoflife/GameOfLifeComponent.java]: I fail you for static variables.

File: src/main/java/lesser/gameoflife/GameOfLifeFrame.java [hunk 2]
```diff
@@ -24,21 +28,62 @@ public GameOfLifeFrame() {
...
+                "Choose an option to load the RLE pattern",
+                "Load RLE Pattern",
+                JOptionPane.DEFAULT_OPTION,
+                JOptionPane.INFORMATION_MESSAGE,
+                null, options, options[0]);
 
+        if (choice == 0) {
+            // Load from URL
+            try {
+                URL url = new URL("https://conwaylife.com/patterns/glider.rle");
```
Comment (andrewoid) [inline src/main/java/lesser/gameoflife/GameOfLifeFrame.java]: If you're loading from a URL, allow the person to choose the URL.

File: src/main/java/lesser/gameoflife/GameOfLifeRleParser.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class GameOfLifeRleParser {
+
+    public static void loadPatternFromFile(GameOfLife game, String filePath) throws IOException {
```
Comment (andrewoid) [inline src/main/java/lesser/gameoflife/GameOfLifeRleParser.java]: I fail you for static methods.

File: src/main/java/lesser/gameoflife/GameOfLifeRleParser.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    }
+
+    public static void loadPatternFromReader(GameOfLife game, BufferedReader reader) throws IOException {
+        String line;
+        int x = 0;
+        int y = 0;
+        boolean inPattern = false;
+
+        Pattern patternSize = Pattern.compile("x\\s*=\\s*(\\d+),\\s*y\\s*=\\s*(\\d+)");
```
Comment (andrewoid) [inline src/main/java/lesser/gameoflife/GameOfLifeRleParser.java:29]: What does this line do?

File: src/test/java/lesser/gameoflife/GameofLifeTest.java [hunk 5]
```diff
@@ -52,4 +57,34 @@ public void setCell() {
 
         assertEquals(1, grid[1][1], "Cell (1,1) should be set to alive");
     }
+
+    @Test
+    void loadGliderPatternFromUrl() throws IOException {
```
Comment (andrewoid) [inline src/test/java/lesser/gameoflife/GameofLifeTest.java]: This is testing loadPatternFromReader()

File: src/test/java/lesser/gameoflife/GameofLifeTest.java [hunk 6]
```diff
@@ -52,4 +57,34 @@ public void setCell() {
...
+
+        assertEquals(3, game.getGrid().length, "Grid height should be 3");
+        assertEquals(3, game.getGrid()[0].length, "Grid width should be 3");
+
+        for (int i = 0; i < expectedGrid.length; i++) {
+            for (int j = 0; j < expectedGrid[i].length; j++) {
+                assertEquals(expectedGrid[i][j], game.getGrid()[i][j],
+                        "Cell (" + i + "," + j + ") does not match the expected value.");
+            }
+        }
```
Comment (andrewoid) [inline src/test/java/lesser/gameoflife/GameofLifeTest.java]: Use assertArrayEquals

File: src/test/java/lesser/gameoflife/GameofLifeTest.java [hunk 7]
```diff
@@ -52,4 +57,34 @@ public void setCell() {
...
+
+        // Then
+        int[][] expectedGrid = new int[][]{
+                {0, 1, 0},
+                {0, 0, 1},
+                {1, 1, 1}
+        };
+
+        assertEquals(3, game.getGrid().length, "Grid height should be 3");
+        assertEquals(3, game.getGrid()[0].length, "Grid width should be 3");
```
Comment (andrewoid) [inline src/test/java/lesser/gameoflife/GameofLifeTest.java]: Two lines aren't necessary with below comment.

File: src/main/java/lesser/gameoflife/GameOfLifeRleParser.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                }
+                continue;
+            }
+
+            // Parse the pattern data
+            if (inPattern) {
+                int count = 0;
+                for (char c : line.toCharArray()) {
+                    if (Character.isDigit(c)) {
+                        count = count * 10 + (c - '0');
```
Comment (andrewoid) [inline src/main/java/lesser/gameoflife/GameOfLifeRleParser.java:54]: What does c - '0' do?

File: src/main/java/lesser/gameoflife/GameOfLifeRleParser.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        }
+    }
+
+    public void loadPatternFromReader(BufferedReader reader) throws IOException {
+        String line;
+        int x = 0;
+        int y = 0;
+        boolean inPattern = false;
+
+        Pattern patternSize = Pattern.compile("x\\s*=\\s*(\\d+),\\s*y\\s*=\\s*(\\d+)");
```
Comment (andrewoid) [inline src/main/java/lesser/gameoflife/GameOfLifeRleParser.java:29]: What does this do?

File: src/main/java/lesser/gameoflife/GameOfLifeRleParser.java (hunk 4 above)
Comment (Lauren34) [reply src/main/java/lesser/gameoflife/GameOfLifeRleParser.java:29]: It creates a regular expression pattern used to find and extract the dimensions of the grid from the RLE file in the format x = <number>, y = <number>. Meaning, it extracts the width and height of the grid from the input file when it encounters a line like "x = 50, y = 30".

File: src/main/java/lesser/gameoflife/GameOfLifeRleParser.java (hunk 8 above)
Comment (Lauren34) [reply src/main/java/lesser/gameoflife/GameOfLifeRleParser.java:54]: It converts digit characters from the RLE input into their corresponding integer values which allows for the proper counting of repeated cells in the pattern.

File: src/test/java/lesser/gameoflife/GameofLifeTest.java [hunk 10]
```diff
@@ -52,4 +57,35 @@ public void setCell() {
...
+
+    @Test
+    void loadGliderPatternFromReader() throws IOException {
+        // Given
+        GameOfLife game = new GameOfLife(10, 10);
+        URL url = new URL("https://conwaylife.com/patterns/glider.rle");
+        URLConnection connection = url.openConnection();
+        BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
+
+        GameOfLifeRleParser parser = new GameOfLifeRleParser(game);
```
Comment (andrewoid) [inline src/test/java/lesser/gameoflife/GameofLifeTest.java]: Put this test in `GameOfLifeRleParserTest`

File: src/test/java/lesser/gameoflife/GameOfLifeRleParserTest.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+        // Then
+        assertEquals(3, game.getGrid().length, "Grid height should be 3");  // Adjust grid height check
+        assertEquals(3, game.getGrid()[0].length, "Grid width should be 3");  // Adjust grid width check
+
+        // Check the expected pattern
+        for (int i = 0; i < expectedGrid.length; i++) {
+            assertArrayEquals(expectedGrid[i], game.getGrid()[i], "Row " + i + " does not match the expected values.");
+        }
+    }
```
Comment (andrewoid) [inline src/test/java/lesser/gameoflife/GameOfLifeRleParserTest.java]:
```suggestion
            assertArrayEquals(expectedGrid, game.getGrid());
```

**Review by andrewoid — APPROVED** [review]

**Pull request merged**